src/libraries/System.Private.CoreLib/src/System/IO/Directory.Unix.cs

```diff
@@ -20,5 +23,37 @@ private static DirectoryInfo CreateDirectoryCore(string path, UnixFileMode unixC
 
             return new DirectoryInfo(path, fullPath, isNormalized: true);
         }
+
+        private static unsafe string CreateTempSubdirectoryCore(string? prefix)
+        {
+            // mkdtemp takes a char* and overwrites the XXXXXX with six characters
+            // that'll result in a unique directory name.
+            string tempPath = Path.GetTempPath();
+            int tempPathByteCount = Encoding.UTF8.GetByteCount(tempPath);
+            int prefixByteCount = prefix is not null ? Encoding.UTF8.GetByteCount(prefix) : 0;
+            int totalByteCount = tempPathByteCount + prefixByteCount + 6 + 1;
+
+            Span<byte> path = totalByteCount <= 256 ? stackalloc byte[256].Slice(0, totalByteCount) : new byte[totalByteCount];
+            int pos = Encoding.UTF8.GetBytes(tempPath, path);
+            if (prefix is not null)
+            {
+                pos += Encoding.UTF8.GetBytes(prefix, path.Slice(pos));
+            }
+            path.Slice(pos, 6).Fill((byte)'X');
+            path[pos + 6] = 0;
+
+            // Create the temp directory.
+            fixed (byte* pPath = path)
+            {
+                if (Interop.Sys.MkdTemp(pPath) is null)
+                {
+                    Interop.CheckIo(-1);
```

Comment:
This way of throwing exception for last error is unintuitive to me.

Reply:
I've added a new internal method `ThrowIOExceptionForLastError()` and used it in all places that were doing this.